File: src/lib/scala/Slate.Common/src/main/scala/slate/common/info/Status.scala
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2015 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
  */

package slate.common.info

import slate.common.{TimeSpan, DateTime}

case class Status (
                    started    : DateTime = DateTime.now()    ,
                    ended      : DateTime = DateTime.now()    ,
                    duration   : TimeSpan = TimeSpan(0,0,0)   ,
                    status     : String   = "not-started"     ,
                    errors     : Int      = 0                 ,
                    error      : String   = "n/a"
                  )
{
  def start(statusName:Option[String] = None): Status = {
    copy(started = DateTime.now(), status = statusName.getOrElse("started"))
  }


  def error(msg:String):Status = {
    copy(error = msg, errors = errors + 1)
  }


  def end(statusName:Option[String] = None): Status = {
    copy(started = DateTime.now(), status = statusName.getOrElse("ended"))
  }
}


object Status
{
  val none = new Status()
}

File: src/lib/kotlin/slatekit-common/src/main/kotlin/slatekit/common/info/Status.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.common.info

import slatekit.common.DateTime
import slatekit.common.TimeSpan
import java.time.Duration

data class Status(
        val started: DateTime = DateTime.now(),
        val ended: DateTime = DateTime.now(),
        val duration: Duration = started.durationFrom(ended),
        val status: String = "not-started",
        val errors: Int = 0,
        val error: String = "n/a"
) {
    fun start(statusName: String? = null): Status =
            copy(started = DateTime.now(), status = statusName ?: "started")


    fun error(msg: String): Status = copy(error = msg, errors = errors + 1)


    fun end(statusName: String? = null): Status {
        val now = DateTime.now()
        val duration = started.durationFrom(now)
        return copy(ended = DateTime.now(), status = statusName ?: "ended", duration = duration)
    }


    companion object {
        @JvmStatic
        val none = Status()
    }
}